include/openthread/border_agent.h

```diff
@@ -248,16 +248,19 @@ otError otBorderAgentSetEphemeralKey(otInstance *aInstance,
  * Requires `OPENTHREAD_CONFIG_BORDER_AGENT_EPHEMERAL_KEY_ENABLE`.
  *
  * Can be used to cancel a previously set ephemeral key before it times out. If the Border Agent is not running or
- * there is no ephemeral key in use, calling this function has no effect.
+ * there is no ephemeral key in use, calling this function has no effect and returns `OT_ERROR_NONE`.
  *
  * If a commissioner is connected using the ephemeral key and is currently active, calling this function does not
- * change its state. In this case the `otBorderAgentIsEphemeralKeyActive()` will continue to return `TRUE` until the
- * commissioner disconnects.
+ * change its state, and `OT_ERROR_INVALID_STATE` is returned. In this case the `otBorderAgentIsEphemeralKeyActive()`
+ * will continue to return `TRUE` until the commissioner disconnects.
  *
  * @param[in] aInstance    The OpenThread instance.
  *
+ * @retval OT_ERROR_NONE           Successfully cleared the ephemeral key.
+ * @retval OT_ERROR_INVALID_STATE  Border Agent is connected to an external commissioner.
```

Comment:
Technically, the ephemeral key is no longer valid once a secure session has been established using the ephemeral key. Why is it useful to return an error when a secure session is active?

Reply:
With active secure session the ephemeral key is no longer valid, but in the current border agent logic it's still marked as 'mUsingEphemeralKey = true` and `otBorderAgentIsEphemeralKeyActive` will continue to return true. Also it's not possible to set a new ephemeral key unless the secure session terminates. 

The caller of `otBorderAgentClearEphemeralKey` may think that after the call it's possible to create a new ephemeral key session, which is not always true. Returning an error so caller doesn't have to call another `otBorderAgentIsEphemeralKeyActive` to know if there is still ephemeral key in use.

Reply:
Longer term, we may want the Border Agent to support multiple secure sessions simultaneously, each one established with its own ephemeral key. We should think about what's the best way to support that model.

Reply:
When do we expect the discussion and implementation for multiple epskc & secure session support to happen? (seems to be discussed here https://threadgroup.atlassian.net/browse/SPEC-1295) That seems to be affecting all existing apis for epskc:
```
otBorderAgentSetEphemeralKey
otBorderAgentClearEphemeralKey
otBorderAgentIsEphemeralKeyActive
```
are all designed for single epskc at the moment. Return values for otBorderAgentSetEphemeralKey need to change, otBorderAgentClearEphemeralKey need to specify an ephemeral key, and I expect that we need to correlate generated epskc and the user, so users can manage their own epskc without affecting others. 


For the current api, if user wants to cancel the ephemeral key mode, and they can't do so due to the active connection, we may want to prompt some information to user so they know what's going on (e.g. reboot phone to disconnect). What would you recommend going forward here?

Reply:
> For the current API, if the user wants to cancel the ephemeral key mode, and they can't do so due to the active connection, we may want to prompt some information to the user so they know what's going on (e.g. reboot phone to disconnect). What would you recommend going forward here?

My view is that we provide a bunch of APIs to give control to the next layer so different UX flows can be implemented:

* `otBorderAgentClearEphemeralKey()` will clear a previously set ephemeral key (if any). As Jonathan said, since the ephemeral key is one-time use, once we have a connection, it cannot be used again (it sort of auto-clears after one use). So practically, we can say `ClearEphemeralKey()` is not failing.
* `otBorderAgentIsEphemeralKeyActive()` indicates and tracks which key is being used, PSKc or ephemeral key (if we have a session with an ephemeral key, the PSKc is not going to be used until it disconnects).
* `otBorderAgentGetState()` can be used to determine whether a commissioner or commissioner candidate is connected or not.

It is always possible for `otBorderAgentSetEphemeralKey()` to fail due to a commissioner being connected, irrespective of whether the commissioner used PSKC or an ephemeral key to connect. So in the next layer code and UX flow, we should handle this. To me, this may be a better place to check and handle such an error rather than from `ClearEphemeralKey()`.

One suggestion here is to maybe add new APIs to allow the Border Agent to be started or stopped (e.g. `otBorderAgentStart/Stop()`). We don't expose these as public APIs today. These can be used to force a disconnect of any currently connected commissioner.

Reply:
If we support multiple epskc secure sessions, one epskc having active secure session will not block creating a new epskc, so sounds good to me that ClearEphemeralKey() doesn't return an error. For the current UX flow we can let api callers of `otBorderAgentClearEphemeralKey()` check `otBorderAgentIsEphemeralKeyActive` and decide what they want to do.

> otBorderAgentIsEphemeralKeyActive() indicates and tracks which key is being used, PSKc or ephemeral key (if we have a session with an ephemeral key, the PSKc is not going to be used until it disconnects).

If we are going to support multiple epskc sessions, I wonder if we will also allow both pskc session and epskc sessions?

Reply:
When surfacing this feature to the user, we need to provide the user a way to cancel the credentials sharing operation and terminates any ongoing sessions immediately to the user changes his/her mind. Abtin's suggestion of adding the public otBorderAgentStart()/otBorderAgentStop() APIs works, but it will be simpler to add a boolean parameter to the existing `otBorderAgentClearEphemeralKey()` API. Thoughts? @abtink @jwhui

Reply:
I prefer the `otBorderAgentStart()/otBorderAgentStop()` APIs because they are more general and can terminate a connected Commissioner (candidate) regardless of whether it used PSKc or an ephemeral key.

Additionally, from an API backward compatibility perspective, adding a new API is preferable to changing existing ones.

Reply:
Is it OK interpret the intention of `otBorderAgentClearEphemeralKey()` while there's an active epskc-based session is that the upper layer hopes to abort the current session? Event we support multiple session in future, the number of sessions are limited, and we need an approach for upper layer to abort some existing sessions for new sessions any way. Thoughts?

Reply:
Besides, from the `meshcop` state bitmap definition, it seems the border agent service is supposed to be always available. Do we want to expose APIs for upper layer to start/stop the service?

Reply:
> Is it OK interpret the intention of otBorderAgentClearEphemeralKey() while there's an active epskc-based session is that the upper layer hopes to abort the current session? Event we support multiple session in future, the number of sessions are limited, and we need an approach for upper layer to abort some existing sessions for new sessions any way. Thoughts?

That's what I think the `otBorderAgentClearEphemeralKey()` should do. From API design perspective, the API should allow the user to change his/her mind after activating the key but not indefinitely waiting for the commissioner to disconnect